### PR TITLE
Object-level validation, frontend

### PR DESCRIPTION
**Pull request by fiskus**
* Added one more visible state to files - `"invalid"`
* Moved entries validation to the function used `final-form` API (`<RF.Field validate={...here...} />`)
* Populate `Ajv` error with entry context data, and then check if there are added or existing files with that error

Screenshot from 2023-03-02 14-33-10

- [x] Documentation
- [x] [Changelog](CHANGELOG.md) entry (skip if change is not significant to end users, e.g. docs only)

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1201258662284091

### CODE REVIEW

File: docs/advanced-features/workflows.md [hunk 1]
```diff
@@ -623,6 +629,31 @@ Requires a README
 }
 ```
 
+##### `s3://bucket/must-have-foo-bar-meta.json`
```
Comment (fiskus) [inline docs/advanced-features/workflows.md:632]: @sir-sigurd please, check if the docs, and example in particular, are correct

Comment (sir-sigurd) [reply docs/advanced-features/workflows.md:632]: I guess I need to drop my example in https://github.com/quiltdata/quilt/pull/3286/files, since your doc seems more complete.

File: docs/advanced-features/workflows.md [hunk 2]
```diff
@@ -623,6 +629,31 @@ Requires a README
...
+
+Requires `{ foo: 'bar' }` object as user specified metadata
```
Comment (sir-sigurd) [inline docs/advanced-features/workflows.md:634]: Maybe format this as JSON object?
```suggestion
Requires `{ "foo": "bar" }` object as user specified metadata
```

File: docs/advanced-features/workflows.md [hunk 3]
```diff
@@ -623,6 +629,31 @@ Requires a README
...
+    "properties": {
+      "meta": {
+        "type": "object",
+        "properties": {
+          "foo": {
+            "type": "string",
+            "pattern": "^bar$"
+          }
+        },
+        "required": ["foo"]
```
Comment (sir-sigurd) [inline docs/advanced-features/workflows.md:650]: Just in case you want it to be exactly `{ "foo": "bar" }`, you'd need to set `"additionalProperties": false` or just use `const`.

**Review by sir-sigurd — COMMENTED** [review]
Doc looks OK but might make sense to use `logical_key` in example to demonstrate how to validate specific object instead of just some object in package.

File: catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx [hunk 4]
```diff
@@ -194,7 +195,13 @@ interface DispatchFilesAction {
...
 }
 
-type FilesEntryState = 'deleted' | 'modified' | 'unchanged' | 'hashing' | 'added'
+type FilesEntryState =
+  | 'deleted'
+  | 'modified'
+  | 'unchanged'
+  | 'hashing'
+  | 'added'
+  | 'invalid'
```
Comment (nl0) [inline catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx:204]: to me "invalid" property seems orthogonal to the state

Comment (nl0) [reply catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx:204]: since, for example, both a modified and an added entry can be invalid in the same time

**Review by nl0 — APPROVED** [review]
looks sane.
i'm not quite sure if you're still showing the validation errors "not assigned" to any entries (or are they not expected to be present at all times?). e.g. if you have some error related to several fields or smth like that. also, you only get (`.find()`) the first error matching entry path, but can there be more than one per path and they are not shown?

File: catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx (hunk 4 above)
Comment (fiskus) [reply catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx:204]: I agree, and it was my first attempt to implement invalid micro-state for any other state (added/existing), but it was too difficult

Now, I think it's not that hard (probably, because I forgot complexities), and I would like to fix it as a tech/UX dept in a separate task

**Comment by fiskus** [comment]
> i'm not quite sure if you're still showing the validation errors "not assigned" to any entries 

Yes, all entries validation errors are below FilesInput. Unfortunately, they are not very helpful, if you don't know exact JSON Schema. At the same time, I don't show error message in entry. There is a room for improvement here, I address it in the next task.

Probably, it's not so important to invest a lot of time into entries meta validation on frontend

> you only get (.find()) the first error matching entry path, but can there be more than one per path and they are not shown?

With `.find` I only find the fact, the entry is invalid. Actual errors are below the FilesInput block.

**Comment by fiskus** [comment]
This is not very good UX, of course. For example, if Schema requires `README.md` with meta, but user provide two files `foo.txt` and `bar.txt`, then both files are invalid with two errors each (1. this is not README.md, this doesn't have meta). But actually, files are not invalid, user just need one more file.

File: docs/advanced-features/workflows.md (hunk 3 above)
Comment (fiskus) [reply docs/advanced-features/workflows.md:654]: Got it, but my intention was to have any fields AND require `{ "foo": "bar" }`

**Comment by fiskus** [comment]
Updated docs: added `logical_key`

**Review by nl0 — APPROVED** [review]

**Review by sir-sigurd — APPROVED** [review]
